src/Polysemy/Scoped.hs

```diff
@@ -90,103 +95,112 @@ interpretScopedAs resource =
   interpretScoped \ p use -> use =<< resource p
 {-# inline interpretScopedAs #-}
 
--- |Higher-order interpreter for 'Scoped' that allows the handler to use additional effects that are interpreted by the
--- resource allocator.
+-- | Higher-order interpreter for 'Scoped' that allows the handler to use
+-- additional effects that are interpreted by the resource allocator.
 --
--- /Note/: It is necessary to specify the list of local interpreters with a type application; GHC won't be able to
--- figure them out from the type of @withResource@:
+-- /Note/: It is necessary to specify the list of local interpreters with a type
+-- application; GHC won't be able to figure them out from the type of
+-- @withResource@:
 --
 -- > interpretScopedWithH @[State Bool, Reader Int] withResource \ _ -> \case
 -- >   SomeAction -> put . (> 0) =<< ask
 -- > where
 -- >   withResource param use = runState False (runReader 5 (use ()))
 interpretScopedWithH ::
   ∀ extra param resource effect r r1 .
-  r1 ~ (Append extra r) =>
-  InsertAtIndex 0 '[] r1 r r1 extra =>
-  InsertAtIndex 1 '[Scoped param resource effect] r1 r (Scoped param resource effect : r1) extra =>
+  (KnownList extra, r1 ~ Append extra r) =>
   (∀ x . param -> (resource -> Sem r1 x) -> Sem r x) ->
   (∀ r0 x . resource -> effect (Sem r0) x -> Tactical effect (Sem r0) r1 x) ->
   InterpreterFor (Scoped param resource effect) r
 interpretScopedWithH withResource scopedHandler =
   interpretWeaving \case
     Weaving (InScope param main) s wv ex _ ->
-      ex <$> withResource param \ resource -> inScope (insertAt @1 @extra (wv (main resource <$ s)))
+      ex <$> withResource param \ resource -> inScope $
+        restack
+          (injectMembership
+           (singList @'[Scoped param resource effect])
+           (singList @extra)) $ wv (main resource <$ s)
     _ ->
       error "top level Run"
   where
     inScope :: InterpreterFor (Scoped param resource effect) r1
     inScope =
       interpretWeaving \case
         Weaving (InScope param main) s wv ex _ ->
-          insertAt @0 @extra (ex <$> withResource param \ resource -> inScope (wv (main resource <$ s)))
+          restack (extendMembershipLeft (singList @extra))
+            (ex <$> withResource param \resource ->
+                inScope (wv (main resource <$ s)))
         Weaving (Run resource act) s wv ex ins ->
-          ex <$> runTactics s (raise . inScope . wv) ins (inScope . wv) (scopedHandler resource act)
+          -- Smells dubious, ask. God I hate tactics.
```

Comment:
! @tek

Reply:
what do you suspect to go wrong there?

Reply:
Oh yeah, uhh, figured I'd ask this in a call. I don't think I've ever had to use `runTactics` outside of `interpretH`, but that could be when I've worked on `Exceptional` I've based the work on the core rewrite. If this is just code adapted from `interpretH`, then it's fine; otherwise, it seems dubious.

Reply:
hm yeah I think I just copied that from `interpretH`